File: worker/scripts/analyze-lead-website-structure.py
#!/usr/bin/env python3
"""Scan lead websites and infer section blueprints per trade bucket."""
from __future__ import annotations

import json
import re
import sys
import urllib.error
import urllib.request
from collections import Counter, defaultdict
from concurrent.futures import ThreadPoolExecutor, as_completed
from csv import DictReader
from datetime import datetime, timezone
from pathlib import Path

CSV_PATH = Path(r"e:\MasterVault\Moonrise\LeadFinderCloud\data\all-leads.csv")
OUT_PATH = Path(r"e:\MasterVault\Moonrise\moonrise-studio\worker\scripts\lead-structure-analysis.json")
LIMIT = int(sys.argv[1]) if len(sys.argv) > 1 else 350
CONCURRENCY = 10
TIMEOUT = 12

TRADE_BUCKETS = [
    ("barber_salon", re.compile(r"barber|hair salon|beauty salon|nail salon|spa|massage", re.I)),
    ("dental_medical", re.compile(r"dental|dentist|orthodont|chiropract|physical therapy|clinic|doctor|medical|urgent care|optomet|dermatolog|pediatric", re.I)),
    ("restaurant_food", re.compile(r"restaurant|cafe|coffee|bakery|pizza|food|catering|juice|\bbar\b|grill|kitchen|diner", re.I)),
    ("fitness", re.compile(r"gym|fitness|yoga|pilates|crossfit|martial|training|sport", re.I)),
    ("home_services", re.compile(r"plumb|electric|hvac|roof|landscap|lawn|pest|clean|paint|handyman|pressure wash|pool|flooring|carpet|tree service|contractor|construction|garage door", re.I)),
    ("auto", re.compile(r"auto repair|car detail|tire|oil change|body shop|towing|mechanic", re.I)),
    ("legal_finance", re.compile(r"attorney|law firm|lawyer|legal|accounting|insurance|financial|real estate", re.I)),
    ("pet", re.compile(r"veterinar|pet groom|dog train|pet board|animal", re.I)),
    ("professional", re.compile(r"marketing|agency|consult|design|photograph|software|tech", re.I)),
]
DEFAULT_BUCKET = "local_service"

SECTION_DETECTORS: dict[str, re.Pattern[str]] = {
    "navigation": re.compile(r"\b(nav|navbar|site-header|main-menu|menu-toggle)\b|<nav[\s>]", re.I),
    "hero": re.compile(r"\b(hero|banner|jumbotron|masthead|page-header)\b|class=\"[^\"]*hero", re.I),
    "credibility": re.compile(r"\b(licensed|insured|certified|accredited|years of experience|since 19|since 20|\b\d+\+?\s*years\b|bbb|award|trusted by|as seen on)\b", re.I),
    "services": re.compile(r"\b(our services|services we offer|what we do|service area|service menu)\b|\bservices\b", re.I),
    "about": re.compile(r"\b(about us|who we are|our story|meet the team|our team|our mission)\b", re.I),
    "gallery": re.compile(r"\b(gallery|portfolio|our work|before and after|photo gallery|project gallery)\b", re.I),
    "testimonials": re.compile(r"\b(testimonial|what (our )?clients say|customer reviews|google reviews|review carousel|client stories)\b|\breviews\b", re.I),
    "pricing": re.compile(r"\b(pricing|our rates|price list|packages|plans)\b", re.I),
    "faq": re.compile(r"\b(faq|frequently asked)\b", re.I),
    "team": re.compile(r"\b(meet (the )?team|our (staff|dentists|doctors|barbers|stylists|trainers|attorneys))\b", re.I),
    "hours_location": re.compile(r"\b(hours of operation|business hours|find us|visit us|directions|service area)\b|\b(monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b.*\b(am|pm|closed)\b", re.I),
    "map": re.compile(r"\b(google\.com/maps|maps\.google|iframe[^>]+maps|data-map|mapbox)\b", re.I),
    "contact_form": re.compile(r"<form[\s\S]{0,2500}?(name|phone|email|message|submit)", re.I),
    "cta_band": re.compile(r"\b(book (now|online|appointment)|schedule (a )?(consultation|appointment)|get (a )?quote|call now|contact us today|free estimate)\b", re.I),
    "footer": re.compile(r"<footer[\s>]", re.I),
}

CANONICAL_ORDER = [
    "navigation",
    "hero",
    "credibility",
    "services",
    "about",
    "gallery",
    "testimonials",
    "pricing",
    "team",
    "faq",
    "hours_location",
    "map",
    "cta_band",
    "contact_form",
    "footer",
]

SECTION_LABELS = {
    "navigation": "Navigation",
    "hero": "Hero",
    "credibility": "Credibility / trust strip",
    "services": "Services",
    "about": "About",
    "gallery": "Photo gallery / portfolio",
    "testimonials": "Testimonials / reviews",
    "pricing": "Pricing / packages",
    "team": "Team / staff",
    "faq": "FAQ",
    "hours_location": "Hours & location",
    "map": "Map embed",
    "cta_band": "CTA band",
    "contact_form": "Contact form",
    "footer": "Footer",
}


def trade_bucket(category: str, name: str) -> str:
    hay = f"{category} {name}"
    for bid, pat in TRADE_BUCKETS:
        if pat.search(hay):
            return bid
    return DEFAULT_BUCKET


def valid_site(url: str) -> bool:
    u = (url or "").strip()
    if not re.match(r"^https?://", u, re.I):
        return False
    if re.search(r"google\.(com|ca)|gstatic\.com|facebook\.com/pg|instagram\.com|yelp\.com|bbb\.org", u, re.I):
        return False
    return True


def fetch_html(url: str) -> str:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "MoonriseStudioStructureBot/1.0",
            "Accept": "text/html,application/xhtml+xml",
        },
    )
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as res:
            ct = res.headers.get("Content-Type", "")
            if "html" not in ct.lower() and "text" not in ct.lower():
                return ""
            data = res.read(1_500_000)
            return data.decode("utf-8", errors="replace")
    except (urllib.error.URLError, TimeoutError, ValueError):
        return ""


def detect_sections(html: str) -> dict[str, bool]:
    return {k: bool(p.search(html)) for k, p in SECTION_DETECTORS.items()}


def order_sections(sections: dict[str, bool]) -> list[str]:
    return [s for s in CANONICAL_ORDER if sections.get(s)]


def pct(n: int, d: int) -> int:
    return round((n / d) * 100) if d else 0


def typical_order(samples: list[list[str]]) -> list[str]:
    pos: dict[str, list[int]] = defaultdict(list)
    counts: Counter[str] = Counter()
    for order in samples:
        for idx, sec in enumerate(order):
            pos[sec].append(idx)
            counts[sec] += 1
    min_hits = max(3, int(len(samples) * 0.2))
    ranked = [s for s, c in counts.items() if c >= min_hits]
    return sorted(ranked, key=lambda s: sum(pos[s]) / len(pos[s]))


def build_blueprint(scanned: int, hits: Counter[str], order_samples: list[list[str]]) -> dict:
    freq = {s: pct(hits[s], scanned) for s in SECTION_DETECTORS}
    ranked = sorted(freq.items(), key=lambda x: -x[1])
    core = [s for s, r in ranked if r >= 55]
    common = [s for s, r in ranked if 30 <= r < 55]
    optional = [s for s, r in ranked if 15 <= r < 30]
    if not core:
        core = [s for s, _ in ranked[:6]]
    return {
        "sampleSize": scanned,
        "coreSections": core,
        "commonSections": common,
        "optionalSections": optional,
        "frequencies": freq,
        "typicalOrder": typical_order(order_samples),
        "structure": [SECTION_LABELS.get(s, s) for s in typical_order(order_samples) or core],
    }


def main() -> None:
    if not CSV_PATH.exists():
        raise SystemExit(f"Missing CSV: {CSV_PATH}")

    with CSV_PATH.open("r", encoding="utf-8", errors="replace", newline="") as f:
        rows = list(DictReader(f))

    candidates = []
    for row in rows:
        url = (row.get("website_url") or row.get("website") or "").strip()
        if not valid_site(url):
            continue
        category = (row.get("category_group") or row.get("category") or "").strip()
        name = (row.get("business_name") or "").strip()
        candidates.append(
            {
                "url": url,
                "category": category,
                "name": name,
                "bucket": trade_bucket(category, name),
            }
        )

    by_bucket: dict[str, list[dict]] = defaultdict(list)
    for c in candidates:
        by_bucket[c["bucket"]].append(c)

    per_bucket = max(12, (LIMIT + len(by_bucket) - 1) // len(by_bucket))
    sample: list[dict] = []
    for lst in by_bucket.values():
        sample.extend(lst[:per_bucket])
        if len(sample) >= LIMIT:
            break
    sample = sample[:LIMIT]

    print(f"Scanning {len(sample)} sites from {len(candidates)} with websites across {len(by_bucket)} buckets...")

    bucket_hits: dict[str, Counter[str]] = defaultdict(Counter)
    bucket_orders: dict[str, list[list[str]]] = defaultdict(list)
    bucket_scanned: Counter[str] = Counter()

    def scan_one(item: dict) -> tuple[str, dict[str, bool] | None]:
        html = fetch_html(item["url"])
        if not html or len(html) < 400:
            return item["bucket"], None
        sections = detect_sections(html)
        return item["bucket"], sections

    done = 0
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as pool:
        futures = {pool.submit(scan_one, item): item for item in sample}
        for fut in as_completed(futures):
            done += 1
            if done % 25 == 0:
                print(f"  {done}/{len(sample)}")
            try:
                bucket, sections = fut.result()
            except Exception:
                continue
            if not sections:
                continue
            order = order_sections(sections)
            bucket_scanned[bucket] += 1
            for k, v in sections.items():
                if v:
                    bucket_hits[bucket][k] += 1
            if order:
                bucket_orders[bucket].append(order)

    blueprints = {
        bucket: build_blueprint(
            bucket_scanned[bucket],
            bucket_hits[bucket],
            bucket_orders[bucket],
        )
        for bucket in sorted(set(list(by_bucket.keys()) + [DEFAULT_BUCKET]))
        if bucket_scanned[bucket] > 0
    }

    report = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "csvPath": str(CSV_PATH),
        "totalWithWebsite": len(candidates),
        "scanned": sum(bucket_scanned.values()),
        "blueprints": blueprints,
    }
    OUT_PATH.write_text(json.dumps(report, indent=2), encoding="utf-8")
    print("Wrote", OUT_PATH)
    for bucket, bp in blueprints.items():
        print(f"\n{bucket} (n={bp['sampleSize']})")
        print("  →", " → ".join(bp["structure"]))


if __name__ == "__main__":
    main()

File: worker/scripts/analyze-lead-website-structure.js
/**
 * One-off: scan leads with websites and infer common page sections per trade.
 * Usage: node scripts/analyze-lead-website-structure.js [--limit=400] [--concurrency=8]
 */
"use strict";

const fs = require("fs");
const path = require("path");
const { parse } = require("csv-parse/sync");

const CSV_PATH = path.join(__dirname, "..", "..", "..", "LeadFinderCloud", "data", "all-leads.csv");
const OUT_PATH = path.join(__dirname, "lead-structure-analysis.json");

const args = Object.fromEntries(
  process.argv.slice(2).map((a) => {
    const [k, v] = a.replace(/^--/, "").split("=");
    return [k, v ?? "1"];
  })
);
const LIMIT = Number(args.limit || 350);
const CONCURRENCY = Math.min(12, Math.max(2, Number(args.concurrency || 8)));
const FETCH_TIMEOUT_MS = 12000;

/** Broad trade buckets for aggregation. */
const TRADE_BUCKETS = [
  { id: "barber_salon", match: /barber|hair salon|beauty salon|nail salon|spa|massage/i },
  { id: "dental_medical", match: /dental|dentist|orthodont|chiropract|physical therapy|clinic|doctor|medical|urgent care|optomet|dermatolog|pediatric/i },
  { id: "restaurant_food", match: /restaurant|cafe|coffee|bakery|pizza|food|catering|juice|bar\b|grill|kitchen|diner/i },
  { id: "fitness", match: /gym|fitness|yoga|pilates|crossfit|martial|training|sport/i },
  { id: "home_services", match: /plumb|electric|hvac|roof|landscap|lawn|pest|clean|paint|handyman|pressure wash|pool|flooring|carpet|tree service|contractor|construction|garage door/i },
  { id: "auto", match: /auto repair|car detail|tire|oil change|body shop|towing|mechanic/i },
  { id: "legal_finance", match: /attorney|law firm|lawyer|legal|accounting|insurance|financial|real estate/i },
  { id: "pet", match: /veterinar|pet groom|dog train|pet board|animal/i },
  { id: "professional", match: /marketing|agency|consult|design|photograph|software|tech/i },
];

const DEFAULT_BUCKET = "local_service";

function tradeBucket(category, name) {
  const hay = `${category || ""} ${name || ""}`;
  for (const b of TRADE_BUCKETS) {
    if (b.match.test(hay)) return b.id;
  }
  return DEFAULT_BUCKET;
}

function isValidSite(url) {
  const u = String(url || "").trim();
  if (!/^https?:\/\//i.test(u)) return false;
  if (/google\.(com|ca)|gstatic\.com|facebook\.com\/pg|instagram\.com|yelp\.com|bbb\.org/i.test(u)) {
    return false;
  }
  return true;
}

/** Section detectors - keyword/heuristic on lowercased HTML text + attrs. */
const SECTION_DETECTORS = {
  navigation: /\b(nav|navbar|site-header|main-menu|menu-toggle)\b|<nav[\s>]/i,
  hero: /\b(hero|banner|jumbotron|masthead|page-header)\b|class="[^"]*hero/i,
  credibility: /\b(licensed|insured|certified|accredited|years of experience|since 19|since 20|\b\d+\+?\s*years\b|bbb|award|trusted by|as seen on)\b/i,
  services: /\b(our services|services we offer|what we do|service area|service menu)\b|\bservices\b/i,
  about: /\b(about us|who we are|our story|meet the team|our team|our mission)\b/i,
  gallery: /\b(gallery|portfolio|our work|before and after|photo gallery|project gallery)\b/i,
  testimonials: /\b(testimonial|what (our )?clients say|customer reviews|google reviews|review carousel|client stories)\b|\breviews\b/i,
  pricing: /\b(pricing|our rates|price list|packages|plans)\b/i,
  faq: /\b(faq|frequently asked)\b/i,
  team: /\b(meet (the )?team|our (staff|dentists|doctors|barbers|stylists|trainers|attorneys))\b/i,
  hours_location: /\b(hours of operation|business hours|find us|visit us|directions|service area)\b|\b(monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b.*\b(am|pm|closed)\b/i,
  map: /\b(google\.com\/maps|maps\.google|iframe[^>]+maps|data-map|mapbox)\b/i,
  contact_form: /<form[\s\S]{0,2500}?(name|phone|email|message|submit)/i,
  cta_band: /\b(book (now|online|appointment)|schedule (a )?(consultation|appointment)|get (a )?quote|call now|contact us today|free estimate)\b/i,
  footer: /<footer[\s>]/i,
};

async function fetchHtml(url) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), FETCH_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      signal: ctrl.signal,
      headers: {
        "User-Agent": "MoonriseStudioStructureBot/1.0 (+research)",
        Accept: "text/html,application/xhtml+xml",
      },
      redirect: "follow",
    });
    if (!res.ok) return "";
    const ct = res.headers.get("content-type") || "";
    if (!/text\/html|application\/xhtml/i.test(ct) && !ct.includes("text")) return "";
    const buf = await res.arrayBuffer();
    if (buf.byteLength > 1_500_000) return "";
    return new TextDecoder("utf-8", { fatal: false }).decode(buf);
  } catch (_) {
    return "";
  } finally {
    clearTimeout(timer);
  }
}

function detectSections(html) {
  const text = String(html || "");
  const found = {};
  for (const [key, re] of Object.entries(SECTION_DETECTORS)) {
    found[key] = re.test(text);
  }
  return found;
}

function orderFromSections(sections) {
  const canonical = [
    "navigation",
    "hero",
    "credibility",
    "services",
    "about",
    "gallery",
    "testimonials",
    "pricing",
    "team",
    "faq",
    "hours_location",
    "map",
    "cta_band",
    "contact_form",
    "footer",
  ];
  return canonical.filter((k) => sections[k]);
}

async function mapPool(items, fn, concurrency) {
  const out = [];
  let i = 0;
  async function worker() {
    while (i < items.length) {
      const idx = i++;
      out[idx] = await fn(items[idx], idx);
    }
  }
  await Promise.all(Array.from({ length: concurrency }, worker));
  return out;
}

function pct(n, d) {
  return d ? Math.round((n / d) * 100) : 0;
}

function buildBlueprint(bucketStats) {
  const blueprints = {};
  for (const [bucket, data] of Object.entries(bucketStats)) {
    const total = data.scanned || 0;
    if (!total) continue;
    const freq = {};
    for (const [section, count] of Object.entries(data.sectionHits)) {
      freq[section] = pct(count, total);
    }
  const ranked = Object.entries(freq)
      .sort((a, b) => b[1] - a[1])
      .map(([section, rate]) => ({ section, rate }));

    const core = ranked.filter((r) => r.rate >= 55).map((r) => r.section);
    const common = ranked.filter((r) => r.rate >= 30 && r.rate < 55).map((r) => r.section);
    const optional = ranked.filter((r) => r.rate >= 15 && r.rate < 30).map((r) => r.section);

    blueprints[bucket] = {
      sampleSize: total,
      coreSections: core.length ? core : ranked.slice(0, 6).map((r) => r.section),
      commonSections: common,
      optionalSections: optional,
      frequencies: freq,
      typicalOrder: data.typicalOrder || [],
    };
  }
  return blueprints;
}

function typicalOrderFromSamples(samples) {
  const pos = {};
  const counts = {};
  for (const order of samples) {
    order.forEach((sec, idx) => {
      if (!pos[sec]) pos[sec] = [];
      pos[sec].push(idx);
      counts[sec] = (counts[sec] || 0) + 1;
    });
  }
  return Object.keys(pos)
    .filter((k) => (counts[k] || 0) >= Math.max(3, Math.floor(samples.length * 0.2)))
    .sort((a, b) => {
      const avgA = pos[a].reduce((s, v) => s + v, 0) / pos[a].length;
      const avgB = pos[b].reduce((s, v) => s + v, 0) / pos[b].length;
      return avgA - avgB;
    });
}

async function main() {
  if (!fs.existsSync(CSV_PATH)) {
    console.error("CSV not found:", CSV_PATH);
    process.exit(1);
  }
  const raw = fs.readFileSync(CSV_PATH, "utf8");
  const rows = parse(raw, { columns: true, skip_empty_lines: true, relax_column_count: true });

  const candidates = [];
  for (const row of rows) {
    const url = String(row.website_url || row.website || "").trim();
    if (!isValidSite(url)) continue;
    const category = String(row.category_group || row.category || "").trim();
    const name = String(row.business_name || "").trim();
    const bucket = tradeBucket(category, name);
    candidates.push({ url, category, name, bucket });
  }

  // Stratified sample: up to LIMIT, spread across buckets
  const byBucket = new Map();
  for (const c of candidates) {
    if (!byBucket.has(c.bucket)) byBucket.set(c.bucket, []);
    byBucket.get(c.bucket).push(c);
  }
  const perBucket = Math.max(12, Math.ceil(LIMIT / byBucket.size));
  const sample = [];
  for (const list of byBucket.values()) {
    for (let i = 0; i < Math.min(perBucket, list.length); i += 1) {
      sample.push(list[i]);
    }
    if (sample.length >= LIMIT) break;
  }

  console.log(`Scanning ${sample.length} websites across ${byBucket.size} trade buckets...`);

  const bucketStats = {};
  for (const b of [...byBucket.keys(), DEFAULT_BUCKET]) {
    bucketStats[b] = { scanned: 0, sectionHits: {}, orderSamples: [] };
  }

  let done = 0;
  await mapPool(
    sample,
    async (item) => {
      const html = await fetchHtml(item.url);
      done += 1;
      if (done % 25 === 0) console.log(`  ${done}/${sample.length}`);
      if (!html || html.length < 400) return null;
      const sections = detectSections(html);
      const order = orderFromSections(sections);
      const stat = bucketStats[item.bucket] || bucketStats[DEFAULT_BUCKET];
      stat.scanned += 1;
      for (const [k, v] of Object.entries(sections)) {
        if (v) stat.sectionHits[k] = (stat.sectionHits[k] || 0) + 1;
      }
      if (order.length) stat.orderSamples.push(order);
      return { ...item, sections, order };
    },
    CONCURRENCY
  );

  for (const stat of Object.values(bucketStats)) {
    stat.typicalOrder = typicalOrderFromSamples(stat.orderSamples || []);
    delete stat.orderSamples;
  }

  const blueprints = buildBlueprint(bucketStats);
  const report = {
    generatedAt: new Date().toISOString(),
    csvPath: CSV_PATH,
    totalWithWebsite: candidates.length,
    scanned: sample.length,
    bucketStats,
    blueprints,
  };
  fs.writeFileSync(OUT_PATH, JSON.stringify(report, null, 2));
  console.log("Wrote", OUT_PATH);
  for (const [bucket, bp] of Object.entries(blueprints)) {
    console.log(`\n${bucket} (n=${bp.sampleSize})`);
    console.log("  core:", bp.coreSections.join(" → "));
    if (bp.commonSections.length) console.log("  common:", bp.commonSections.join(", "));
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
